### PR TITLE
Implement OAuth2/OpenID based/inspired login flows

### PR DESCRIPTION
**Pull request by matthiasblaesing**
- Implemented frontend and backend implementation for OAuth2 Code flow (or similar implementations)
- Added sample configuration for the currently supported identity  providers: GitHub, Google and Amazon.
- Don't use email as user id, but introduce user table, that maps authentication to email/name 
- Add bootstrap configuration, that does not swallow exceptions, but logs them to the Apache error log
- Updated composer.json to project data and not generic ZendFrame work information
- Removed google sign-in library
- Removed image files copied from ZendFramework (unsed)

### CODE REVIEW

**Comment by matthiasblaesing** [comment]
@jpirek @jkovalsky it would be great if you could have a look at this. This is the PHP implementation of what I prototyped in java and announced on the mailing list in the mail "[PluginPortal] Implement OAuth2 based authentication in PP3?".

I still have a bad feeling using PHP here (see the hack, that was needed to get some sensible error reporting in Module.php), but I think this should work. In fact, I have an experimental setup at:

https://doppel-helix.eu/pp3/

Please note: when this is merged, the database structure needs to be updated and a migration path will be ugly (you can't query the OAuth providers for the userids, you'd need).

File: pp3/composer.json [hunk 1]
```diff
@@ -1,23 +1,22 @@
 {
-    "name": "zendframework/skeleton-application",
-    "description": "Skeleton Application for ZF2",
+    "name": "apache/netbeans-pp3",
+    "description": "Netbeans Plugin Portal 3",
```
Comment (jkovalsky) [inline pp3/composer.json]: Upper case letter B please i.e.: _"NetBeans Plugin Portal 3"_

File: pp3/module/Application/config/module.config.php.dist [hunk 2]
```diff
@@ -15,13 +15,34 @@ return array(
         'mavenRepoUrl' => 'https://repo1.maven.org/maven2/',
         'catalogSavepath' => '/home/honza/checkout/pp3/public/data',
         'catalogUrlPath' => 'http://localhost/checkout/pp3/public/data',    
-        'dtdPath' => 'http://localhost/checkout/pp3/public/data/autoupdate-catalog-2_6.dtd',    
-        'admin' => array(
-            'jan.pirek@gmail.com',
-            'jiri.kovalsky@gmail.com'
```
Comment (jkovalsky) [inline pp3/module/Application/config/module.config.php.dist]: Where do we set initial list of administrators?

**Comment by jkovalsky** [comment]
When trying your solution with Google it results with _NotNullConstraintViolationException_ but in the response I see my name and _@gmail.com_ e-mail address received correctly so I believe it's just missing to set the true/false values for admin/verifier parameters. Can you fix/mock it?

BTW, so far reviewed 6 out of 41 changed files.

File: pp3/module/Application/src/Application/Controller/PluginController.php [hunk 3]
```diff
@@ -349,12 +350,13 @@ private function _sendApprovalEmail($plugin) {
...
+        foreach($this->_userRepository->findVerifier() as $verifier) {
             if ($i == 0) {
-                $mail->addTo($verifier->getUserId());
+                // @todo: get an explanation why the first verifier is special cased
+                $mail->addTo($verifier->getEmail());
                 $i++;
             } else {
-                $mail->addBcc($verifier->getUserId());
+                $mail->addBcc($verifier->getEmail());
             }
```
Comment (jkovalsky) [inline pp3/module/Application/src/Application/Controller/PluginController.php]: @jpirek Why don't we simply add all verifiers on BCC: list?

Comment (jpirek) [reply pp3/module/Application/src/Application/Controller/PluginController.php]: Yes, that makes sence, it will speed things up.

**Comment by matthiasblaesing** [comment]
Thank you for the first review comments. I just pushed an update, that should address the first comments and also survived a first round of "test" verification.

For the how do you become admin: You'll need to have DB access  - the admin and verifier status flags moved as boolean flags to the user. The verifier status can be configured by admin (as before), the admin status always required an admin to fiddle with configuration files, so from my perspective not to much lost there.

**Comment by jpirek** [comment]
Thanks a lot Matthias for all of this work, much apreciated! I am personally totally fine with admin flags in DB. As you said, not big dfference to update config file on disc or run update query on DB...

**Comment by jkovalsky** [comment]
Thanks for fixing the _NotNullConstraintViolationException_ issue. Now I can login with Google fine. Some more comments from me:

- can the Logout link also contain identification of the logged in user (e-mail, id, whatever)?
- trying to login via GitHub account results in 500 Internal Server Error when calling _https://doppel-helix.eu/pp3/login/callback?code=<...>&state=<...>_
- can you please add my jiri.kovalsky@gmail.com e-mail to admins team in the DB so that I can test that functionality?

Thanks again Matthias for your great contribution. And please keep up!

**Comment by matthiasblaesing** [comment]
Thanks for checking again. Concrete feedback:
- Information about the logged in user is not displayed
- the github login failed, because you have no public email associated with your profile - we have permission to read also the non-profile emails, we just need to do it.
- Your account was made administrator + verifier

**Comment by matthiasblaesing** [comment]
One thing before this could go online: Do we need a migration path from the old DB structure to the new one or are we willing to "drop" the current state and let the users rebuild? If we need a migration path I'm willing to look into it, though if I see correctly we have currently 4 plugins, which I would consider droppable.

I have one changeset in the pipeline based on this, that adds license headers and cleans up the JS libraries, which I'll propose as a PR once this is merged.

**Comment by jkovalsky** [comment]
> One thing before this could go online: Do we need a migration path from the old DB structure to the new one or are we willing to "drop" the current state and let the users rebuild? If we need a migration path I'm willing to look into it, though if I see correctly we have currently 4 plugins, which I would consider droppable.

I don't think it's worth the burden. It should be enough to just write down registered plugins and contact to their authors and ask them to republish their plugins once again. After all, they have the most cumbersome part (getting their plugins to Maven Central) already done.

> I have one changeset in the pipeline based on this, that adds license headers and cleans up the JS libraries, which I'll propose as a PR once this is merged.

OK

**Comment by matthiasblaesing** [comment]
@jpirek @jkovalsky thank you for your work and review!

**Pull request merged**